plonky_block_proof_gen/src/verifier_state.rs

```diff
@@ -0,0 +1,110 @@
+//! This module defines the `VerifierState`, that contains the necessary data to
+//! handle succinct block proofs verification.
+
+use std::ops::Range;
+
+use log::info;
+use paste::paste;
+use plonky2_evm::{all_stark::AllStark, config::StarkConfig};
+
+use crate::constants::*;
+use crate::{
+    prover_state::ProverState,
+    types::{AllRecursiveCircuits, VerifierData},
+};
+
+/// Plonky2 verifier state.
+///
+/// The default generation requires generating all the verifier data before
+/// extracting the verifier-related data, which can take a long time and require
+/// a large amount of memory.
+pub struct VerifierState {
+    ///
+    pub state: VerifierData,
+}
+
+/// Builder for the verifier state.
+#[derive(Debug)]
+pub struct VerifierStateBuilder {
+    arithmetic_circuit_size: Range<usize>,
+    byte_packing_circuit_size: Range<usize>,
+    cpu_circuit_size: Range<usize>,
+    keccak_circuit_size: Range<usize>,
+    keccak_sponge_circuit_size: Range<usize>,
+    logic_circuit_size: Range<usize>,
+    memory_circuit_size: Range<usize>,
+}
+
+impl Default for VerifierStateBuilder {
+    fn default() -> Self {
+        // The default ranges are somewhat arbitrary, but should be enough for testing
+        // purposes against most transactions.
+        // Some heavy contract deployments may require bumping these ranges though.
```

Comment:
It's not very user friendly to expect people to know the ranges used for a proof and I think they should be considered internals. It would be better if the ranges used were included in the proof itself (or its metadata) such that it becomes "self-documenting" wrt the parameters used to generate it.

Reply:
You cannot generate proof without specifying preliminary ranges (this is part of the preprocessing).

Reply:
Right, I get that. What I meant was that the proof itself could carry within it the ranges used to generate it. That way a verifier – perhaps one written independently and in COBOL! – wouldn't need to know what ranges were used to generate the proof.

Reply:
Did our offline discussion clear the confusion?

Reply:
I wasn't part of the offline discussion, but aren't the minimum ranges something you can't really know until you attempt to generate the proof?

Reply:
If one wants to generate only the smallest ranges sufficient for a given transaction, yes, this can't be known ahead of time. But once a proof is generated, if you need to change the ranges because you're missing one size for the next proof, you'll need to regenerate prover and verifier data, making the previous ones obsolete and the previously generated proof unverifiable.